### ISSUE TITLE
[FSDP2] MixedPrecisionPolicy reduce_dtype is not applied to the TP-mesh Partial -> Replicate gradient all-reduce

### ISSUE THREAD

**Issue by bogatyy** [pytorch/pytorch#190626]
## Bug description

Under FSDP2 + tensor parallelism (2D `dp x tp` mesh), `MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.float32)` applies `reduce_dtype` only to FSDP's own reduce-scatter. The `Partial() -> Replicate()` all-reduce over the TP mesh, which FSDP itself issues during post-backward for every TP-replicated parameter (e.g. norm weights under sequence parallelism, replicated biases, MoE router weights), runs in the gradient's own dtype, i.e. `param_dtype` bf16. The user asks for an fp32 reduction, FSDP accepts the policy silently, and half of the 2D reduction still happens in bf16.

**Impact: every torchtitan run with `tensor_parallel_degree > 1` on shipped defaults hits this, including the published Llama 3.1 70B and 405B demonstration configs (both tp=8): their norm-weight gradients are TP-reduced in bf16 while the config promises, and only permits, fp32.**

The `MixedPrecisionPolicy` docstring says `reduce_dtype` "specifies the dtype for gradient reduction (i.e. reduce-scatter or all-reduce)" (`torch/distributed/fsdp/_fully_shard/_fsdp_api.py`). The TP all-reduce is an all-reduce that FSDP issues, and it is not covered.

## Minimal repro (4 GPUs, dp=2 x tp=2; also reproduces at 2 GPUs with mesh (1, 2))

```python
# torchrun --standalone --nproc_per_node=4 repro.py
import math
import torch
import torch.nn as nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import fully_shard, MixedPrecisionPolicy
from torch.distributed.tensor import DTensor, Replicate, Shard, distribute_tensor

mesh = init_device_mesh("cuda", (2, 2), mesh_dim_names=("dp", "tp"))
tp = mesh["tp"]
tp_rank = tp.get_local_rank()

class Scale(nn.Module):                       # norm-style TP-Replicate param
    def __init__(self, d):
        super().__init__()
        self.w = nn.Parameter(torch.ones(d))
    def forward(self, x):
        return x * self.w

m = Scale(1024).cuda()
m.w = nn.Parameter(distribute_tensor(m.w.detach(), tp, [Replicate()]))
fully_shard(m, mesh=mesh["dp"], mp_policy=MixedPrecisionPolicy(
    param_dtype=torch.bfloat16, reduce_dtype=torch.float32))

x_local = torch.zeros(8, 1024, device="cuda", dtype=torch.bfloat16)
x_local[0, :] = 1.0 if tp_rank == 0 else 2.0 ** -10   # both exact in bf16
m(DTensor.from_local(x_local, tp, [Shard(0)], run_check=False)).to_local().sum().backward()

g = m.w.grad.full_tensor().double().cpu()[0]
mant = 2.0 * math.frexp(float(g))[0]          # dp SUM/AVG is a power of two
assert mant == 1.0 + 2.0 ** -10, f"TP all-reduce ran in bf16: mantissa {mant}"
```

The per-rank TP partial gradients are exactly `1.0` and `2^-10`, both exact in bf16. Their sum `1 + 2^-10` is exact in fp32 and rounds (no tie) to exactly `1.0` in bf16, whose spacing at 1.0 is `2^-8`. The mantissa of the final (fp32, master-weight) gradient is therefore a bit-exact witness of the collective's dtype. Observed: mantissa `1.0` (the entire low term is lost). Expected with `reduce_dtype=float32`: `1.0009765625`. Deterministic across seeds; identical on torch 2.9.1+cu128 and a 2.12 nightly (2026-04-08) on 4x H100; the mechanism is unchanged on current main.

## Root cause

- `_fsdp_param.py`, `unsharded_grad_data` -> `_get_grad_inner_tensor`: for a DTensor parameter, the gradient is redistributed to the parameter's placements on non-DP mesh dims (`grad.redistribute(...)`). Partial -> Replicate on the TP dim is an all-reduce executed in `grad.dtype` = `param_dtype`. No cast happens anywhere on this path.
- `_fsdp_param_group.py`, `post_backward`: collects `unsharded_grad_data`, i.e. the TP all-reduce has already happened before anything `reduce_dtype`-aware runs.
- `_fsdp_collectives.py`, `foreach_reduce`: `reduce_dtype = reduce_dtype or grad_dtype` and the reduce-scatter buffer allocation are the first and only places `reduce_dtype` takes effect.
- Decisive internal inconsistency: on the gradient-accumulation path, `to_accumulated_grad_if_needed` upcasts the unsharded gradient to `reduce_dtype` before `unsharded_accumulated_grad_data` calls the same `_get_grad_inner_tensor`. So the exact same model reduces the TP dim in fp32 when `set_requires_gradient_sync(False)` was used and in bf16 on the plain path. Two orderings of the same two operations, one of them wrong.

## Reachability

Verified against the shipped torchtitan configs of the era the flagship demonstrations were published (commit db6b39ac, Oct 2024) as well as current main; every trigger condition is a default:

- `train_configs/llama3_70b.toml` and `llama3_405b.toml` both ship `tensor_parallel_degree = 8` with `data_parallel_shard_degree = -1`. (`llama3_8b.toml` ships tp=1 and is unaffected; the trigger is purely tp > 1, model size is irrelevant.)
- `config_manager.py` of that era already locked `--training.mixed_precision_reduce` to `default="float32", choices=["float32"]`; users could not opt out of the fp32-reduce promise then and cannot now.
- `parallelize_llama.py` of that era placed `attention_norm` and `ffn_norm` under `SequenceParallel()`, which replicates the weight over the TP mesh and leaves its gradient Partial, pending the all-reduce that FSDP issues.
- torch v2.5.0 (`_composable/fsdp/_fsdp_param.py:674-682`) carries the same unguarded `Partial -> Replicate` redistribute as current main, so the mechanism spans that era through today. Runtime confirmation is on 2.9.1 and a 2026-04 nightly; the v2.5 evidence is source-level.

The dp size does not matter: the TP all-reduce runs in `_get_grad_inner_tensor` even with a size-1 DP mesh, so single-node tp>1 runs are equally affected.

This is a shipped-default configuration in torchtitan: `mixed_precision_param="bfloat16"` and `mixed_precision_reduce="float32"` (the only permitted value) feed `MixedPrecisionPolicy` directly, `enable_sequence_parallel=True` by default, and the norm weight is placed Replicate on the TP mesh with its pending backward all-reduce explicitly delegated to FSDP (`decoder_sharding.py`: "R if SP (pending BWD AR handled by FSDP)"). Any torchtitan run with `tensor_parallel_degree > 1` and `data_parallel_shard_degree > 1` on defaults reduces its `attention_norm.weight` and `ffn_norm.weight` gradients over the TP group in bf16 while every other parameter is reduced in fp32.

Scope: only TP-Replicate parameters are affected (norm weights, replicated biases, routers); Shard-placed weights have no Partial to reduce. The error is one bf16 rounding of a TP-size-way sum, about 2^-9 relative, not a sign or magnitude corruption. It is exactly the precision the user paid `reduce_dtype=float32` to keep, on the parameters where low-precision reduction is known to matter most, and nothing warns: a user auditing the reduce-scatter buffer dtype sees fp32 and is misled.

## Versions

Verified failing on torch 2.9.1+cu128 and 2.12.0.dev20260408+cu128 (4x H100, NCCL). Mechanism present in current main.

## Related issues (searched, distinct)

- #125191 introduced the accumulation-path upcast (the half that is correct); the plain path never got the same treatment.
- #186998 (reduce_dtype resolving to None after a frozen first forward) and #143297 (clamp reduce_dtype in lazy init) touch dtype resolution, not the DTensor redistribute.

*This report was authored with assistance from Claude; the mechanism, repro, and exact-value oracle were human-reviewed.*

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

**Comment by weifengpy** [comment]
thanks for the detailed description! this is indeed a FSDP2+TP composability problem
* applying TP alone, model.parameters() stays in float32, and reduce dtype is naturally float32
* applying TP + FSDP2 with mixed precision, model.parameters() compute dtype become bf16, and hence DTensor partial-to-replicate become bf16

A more clear contract would be
* TP provide a way to specify reduce dtype
* FSDP2 call partial-to-replicate redistribution and automatically respected reduce dtype

FSDP2 mixed precision policy reduce_dtype was talking about reduce-scatter for 1D mesh, and all-reduce for 2D mesh HSDP. TP grad reduce is still delegated to DTensor redistribution

**Comment by weifengpy** [comment]
chatting with @tianyu-l regarding torchtitan siutation. there was a question on whether this is critical to numerics. @bogatyy did you observe some evidence that numric is sensitive to TP grad reduce being float32, instead of bf16?

**Comment by bogatyy** [comment]
@weifengpy I did not test on large-scale runs, and would not necessarily expect major corruption to happen, because this is literally a rounding error (pun intended).
That said, Llama 3.1 70B and 405B configs do a 8-way reduce, so the precision downgrade matters.
The bug hits norm weights and MoE router weights, which are particularly sensitive to precision.
If you can check either way, please let me know.
Btw is the public torchtitan config/code exactly what was actually used to train Llama 3.1, or a later port of some internal pipeline?

**Comment by weifengpy** [comment]
> The bug hits norm weights and MoE router weights

this was critical to call out. what do you think of fsdp2 being able to specify fp32 for norms compute dtype? TP get fp32 as a result

**Comment by bogatyy** [comment]
@weifengpy Agreed it's a composability issue. One framing point: from the user's perspective this isn't a missing contract, it's an existing one being dropped. `MixedPrecisionPolicy.reduce_dtype` is documented as "the dtype for gradient reduction (reduce-scatter or all-reduce)", and the TP partial-to-replicate is an all-reduce that FSDP itself issues in post-backward, so a user who sets `reduce_dtype=fp32` has already asked for this reduction to be fp32.

I'd lean toward FSDP honoring its existing `reduce_dtype` through to the redistribution (my proposed PR) rather than adding a separate TP-side dtype knob. A new knob would be redundant surface, and it would leave `reduce_dtype` still silently ignored for everyone who already sets it. Whether the upcast lives in FSDP or in a dtype-aware DTensor redistribute is a layering call I'm happy to follow your preference on; the user-facing knob stays `reduce_dtype` either way.

Keeping fp32 norm compute dtype as an opt-in feature sounds good too, but I'd keep it separate from fixing this bug: it changes forward numerics as well, and it wouldn't cover replicated biases or MoE router weights unless those also go fp32.

Separately, still curious on the earlier question: is the public torchtitan Llama 3.1 config/parallelization code the same one used for the actual Llama 3.1 70B/405B training runs, or a later reimplementation of an internal pipeline? It matters for gauging how much real training this touched. CC @tianyu-l

**Comment by weifengpy** [comment]
> MixedPrecisionPolicy.reduce_dtype is documented as "the dtype for gradient reduction (reduce-scatter or all-reduce)"

we can discuss whether this is a fsdp bug or doc inaccuracy. Your PR makes sense if it's a fsdp bug

today the TP + FSDP UX are 2 separate module wrappings
```
tp_model = distribute_module(model, tp_mesh)
fsdp_tp_model = fully_shard(model, tp_model, mp_policy, fsdp_mesh)
```

TP is about sharding on sequence dim: consider TP-only, RMSNorm needs grad reduce base on following placements
* activation is Shard(1) along TP mesh, 1 is sequence dim
* parameter is Replicate
* gradient is Partial
we are all-reduce because of sharding on sequence dim across TP mesh. DTensor redistribution from Partial-to-Replicate is responsible for grad shard

DP is about sharding on batch dim: consider FSDP-only, RMSNorm needs grad reduce base on following placements
* activation is Shard(0) along DP mesh, 0 is batch dim
* parameter is Shard(0)
* gradient is Partial
we are reduce-scatter because of sharding on batch dim along DP mesh

putting DP+TP together, is there a good reason that DP needs to specify how grad should be reduced because of sequence dim sharding?

the only case I am aware of for DP to be responsible for TP's grad reduce is for single-device numerics purpose: https://github.com/pytorch/pytorch/issues/171916 . But your PR is about grad reduce dtype

**Comment by bogatyy** [comment]
@weifengpy Thank you for responding in depth!
I had not considered the possibility of "the code is right, but the docs are wrong", but it makes sense.
I still lean towards interpreting it as a bug, because of the following existing inconsistency.

On the gradient-accumulation path, `to_accumulated_grad_if_needed` upcasts the unsharded grad to `reduce_dtype` before `_get_grad_inner_tensor` runs the same Partial-to-Replicate redistribute. So when `set_requires_gradient_sync(False)` is used, the TP all-reduce runs in fp32, and on the plain path it runs in bf16. With the same model and the same redistribute call, the dtype varies depending only on whether grads were accumulated. The proposed PR would remove this divergence.

This also speaks to your question of whether DP should specify how the TP reduce happens: in one place (this upcast), FSDP already specifies it.
So arguably the intent that `reduce_dtype` covers this reduce is already present, just not applied uniformly.
And it is a dtype case of FSDP owning the TP reduce, distinct from the single-device-numerics case in #171916.
That said, this upcast is about accumulating microbatch grads in higher precision, so the fp32 TP reduce there is a consequence rather than a deliberate choice.

Also, since there is no TP-side reduce-dtype knob today, scoping `reduce_dtype` to exclude the TP reduce would leave a user who set `reduce_dtype=fp32` with no way to get an fp32 reduction for norm and router grads. Putting this effect into the existing knob is the only resolution that makes the setting the user already chose take effect.

Another option to keep the fp32 reduction by default: make the redistribute dtype-aware and have FSDP pass `reduce_dtype` into it. The user-facing knob stays `reduce_dtype` either way.
Happy to reimplement it that way if you prefer.

**Comment by weifengpy** [comment]
> @weifengpy Thank you for responding in depth!

@bogatyy when applying TP-only to bf16 model, no FSDP, what's the plan to support fp32 grad reduce? I assume introducing TP reduce_type knob is the solution?

**Comment by bogatyy** [comment]
@weifengpy I dug around the codebase before addressing your question, and it looks like the mechanism for this already exists, even if the user-facing knob does not. `DTensor.redistribute` takes `forward_dtype`/`backward_dtype` (backward also casts back to the input dtype). Nothing in `torch/distributed` sets `backward_dtype` today, and neither TP nor FSDP passes a dtype into its `redistribute` calls, so a bf16 model reduces norm grads in bf16 on both paths.

To your question, supporting fp32 grad reduce for TP-only: the Partial->Replicate reduce is a backward collective, so `backward_dtype=fp32` is the lever (its cast-back also hands bf16 grads back for the optimizer); the TP reduce_dtype knob you suggested could be a new user-facing param feeding into it.

For FSDP+TP: `_get_grad_inner_tensor` redistributes the grad forward, so `forward_dtype=reduce_dtype`, leaving the result fp32 for the reduce-scatter.

So this PR can pass `forward_dtype=self.reduce_dtype` into the redistribute FSDP already calls, reusing the existing `reduce_dtype` with no new user-facing param. The TP-only user-facing knob can be introduced in a later PR separately.

Caveat: I have not runtime-verified that `forward_dtype` composes as I expect, but if you agree with this direction overall, I am happy to run a check on a distributed build and switch the PR to `forward_dtype`.

**Comment by weifengpy** [comment]
> The TP-only user-facing knob can be introduced in a later PR separately.

suppose you introduce the TP-only reduce dtype knob, when fsdp reduce_dtype
!= TP reduce dtype, which one do you respect in fsdp+tp?

**Comment by bogatyy** [comment]
@weifengpy Good question, and after a bit of digging, the answer seems clear: in FSDP+TP, FSDP's `reduce_dtype` owns the reduce for these grads, with no competing TP reduce to arbitrate against.

The TP-partial grads (norms, biases, routers) come from `Replicate` params on the TP mesh, so their gradient is `Partial`. The TP styles do not reduce that; FSDP does, in post-backward: `_get_grad_inner_tensor` redistributes the grad `Partial -> Replicate` explicitly (this very code path), in `reduce_dtype`. So whoever wraps DP performs it, and here that is FSDP. That also matches mixed precision being an outer-wrapper policy: the layer that performs the reduce owns its dtype.

**Edit:** I found two corrections after initially posting this comment, neither of which changes the fix or the answer above:
1. The PR cannot pass `forward_dtype` into the redistribute instead of upcasting: only the TP-Replicate params hit the redistribute, so `forward_dtype` leaves the Shard params in bf16 and trips `foreach_reduce`'s uniform-dtype assert. I'd keep it upcasting all grads to `reduce_dtype`,  making sure the mixed-precision reduction is a whole-model policy.
2. I earlier suggested a TP `reduce_dtype` knob would govern these grads via `backward_dtype`. It would not: `backward_dtype` governs the activation-gradient reduces at TP boundaries, while these weight grads are never reduced by TP (they stay `Partial` until the DP wrapper reduces them). So there is no TP-vs-FSDP precedence question for them to begin with.

**Comment by bogatyy** [comment]
@weifengpy To close the precedence question: TP itself never issues the reduce for these weight grads (SequenceParallel just wraps them `Replicate` and only hooks the activations, not the gradient), so the `Partial -> Replicate` all-reduce is issued lazily by whoever materializes the grad, which in FSDP+TP is FSDP itself in `_get_grad_inner_tensor`. That makes FSDP the only thing issuing this reduce when the two are composed, so its `reduce_dtype` governs by construction, with no competing TP reduce to arbitrate; a TP-side knob would only take effect in TP-only, and on the activation-gradient reduces that TP does issue.

Does this allow us to converge on the desired behavior in the default case? (DP+TP, no new knobs introduced)
If yes, I see two choices for the right layer to put this: an upcast in FSDP (PR #190627, simpler and already working), or a dtype-aware `redistribute` (trips foreach_reduce's uniform-dtype assert).

**Comment by bogatyy** [comment]
@anshul-si  tagging you since you fixed two recent bugs in this same FSDP2+TP mixed precision seam (#183805, #187376). Curious for your take on the question above: since FSDP is the only thing issuing the Partial -> Replicate all-reduce for these grads, should its `reduce_dtype` govern it (PR #190627), or is there a reason to leave it in compute dtype?
